iocore/eventsystem/unit_tests/test_IOBuffer.cc

```diff
@@ -173,6 +173,168 @@ TEST_CASE("MIOBuffer", "[iocore]")
+
+      // fill all of the current buffer
+      miob->write(buf, 6144);
+      CHECK(miob->max_read_avail() == 8192);
+      CHECK(miob->current_write_avail() == 0);
+      CHECK(miob->high_water() == true);
+      CHECK(miob->current_low_water() == true);
+      CHECK(miob->write_avail() == 0); ///< should have no side effect
+
+      // consume half of the data
+      miob_r->consume(4096);
+      CHECK(miob->max_read_avail() == 4096);
+      CHECK(miob->current_write_avail() == 0);
+      CHECK(miob->high_water() == true);
+      CHECK(miob->current_low_water() == true);
+      CHECK(miob->write_avail() == 0); ///< should have no side effect
+
+      // consume all of the data
+      miob_r->consume(4096);
+      CHECK(miob->max_read_avail() == 0);
+      CHECK(miob->current_write_avail() == 0);
+      CHECK(miob->high_water() == false);
+      CHECK(miob->current_low_water() == true);
+      CHECK(miob->write_avail() == 4096); ///< should have a side effect: add a new block
+
+      CHECK(miob->max_read_avail() == 0);
+      CHECK(miob->current_write_avail() == 4096);
+      CHECK(miob->high_water() == false);
+      CHECK(miob->current_low_water() == false);
+      CHECK(miob->write_avail() == 4096); ///< should have no side effect
+    }
+
+    SECTION("water_mark == block_size()")
+    {
+      miob->water_mark = 4096;
+      REQUIRE(miob->water_mark == miob->block_size());
+
+      // fill half of the current buffer
+      miob->write(buf, 2048);
+      CHECK(miob->max_read_avail() == 2048);
+      CHECK(miob->current_write_avail() == 2048);
+      CHECK(miob->high_water() == false);
+      CHECK(miob->current_low_water() == true);
+      CHECK(miob->write_avail() == 6144); ///< should have a side effect: add a new block
+
+      CHECK(miob->max_read_avail() == 2048);
+      CHECK(miob->current_write_avail() == 6144);
+      CHECK(miob->high_water() == false);
+      CHECK(miob->current_low_water() == false);
+      CHECK(miob->write_avail() == 6144); ///< should have no side effect
+
+      // fill all of the current buffer
+      miob->write(buf, 6144);
+      CHECK(miob->max_read_avail() == 8192);
+      CHECK(miob->current_write_avail() == 0);
+      CHECK(miob->high_water() == true);
+      CHECK(miob->current_low_water() == true);
+      CHECK(miob->write_avail() == 0); ///< should have no side effect
+
+      // consume half of the data
+      miob_r->consume(4096);
+      CHECK(miob->max_read_avail() == 4096);
+      CHECK(miob->current_write_avail() == 0);
+      CHECK(miob->high_water() == false);
+      CHECK(miob->current_low_water() == true);
+      CHECK(miob->write_avail() == 4096); ///< should have a side effect: add a new block
+
+      CHECK(miob->max_read_avail() == 4096);
+      CHECK(miob->current_write_avail() == 4096);
+      CHECK(miob->high_water() == false);
+      CHECK(miob->current_low_water() == true);
+      CHECK(miob->write_avail() == 8192); ///< should have a side effect: add a new block (again!)
```

Comment:
Note: this test fails without `add_block_at_tail()`.

Reply:
Why you add 2 empty IOBufferBlocks at the tail of the `miob`?

The `water_mark` set to 4096 means:
- Once there are 4096 or more bytes of data in `miob`, the SM must be notified to process the data.
- If there are not enough data in `miob`, you must first try to fill it with at least 4096 bytes of data.
- It is required to prepare at least 4096 bytes of space to receive the data.

So the maximum count of blocks is (4096 + block_size - 1 ) / block_size + 1 = (4096 + 4095) / 4096 + 1 = 2.

With `add_block_at_tail()`, you have:
- 4096 bytes of data for read avail which is 1 block
- 8192 bytes of space for write avail which is 2 blocks

Reply:
I think your description of `water_mark` looks the same as my understanding. The problem I'm trying to fix is unnecesary IOBufferBlock allocation. 

Prior to this change, every time `write_avail()` is called it allocates new IOBufferBlock and replaces the next block of `_writer`. Because of adding a new block doesn't change the condition of `current_low_water()` if the `warter_mark` is equivalent to block size.

https://github.com/apache/trafficserver/blob/92d4ef1424f547735bfc33ca847c8ac5ab19cd8a/iocore/eventsystem/I_IOBuffer.h#L1126-L1130

The step-by-step description of the problem is below.

1. let's say MIOBuffer has 1 empty block (size is 4096) & warter_mark is 4096 too 
```
A (0/4096)
```

2. when it's full, `current_write_avail()` becomes 0, which means `current_low_water()` is true
```
A (4096/4096)
```

3. when `write_avail()` is called, it allocates new block and set it next of `_writer`
```
A (4096/4096) -> B (0/4096)
``` 

4. `current_write_avail()` becomes 4096, but `current_low_water()` is still 4096

5. when `write_avail()` is called again, it allocates new block and replace the old empty buffer. The block allocated on step 3. will be freed.
```
A (4096/4096) -> C (0/4096)
``` 

6. step 4. & 5. are repeated everytime `write_avail()` is called.


With the change of this PR, behaviors of later step 5. is different.


5. when `write_avail()` is called again, it allocates new block and append it in the tail of the list
```
A (4096/4096) -> B (0/4096) -> C (0/4096)
```

6. `current_write_avail()` becomes 8192 and `current_low_water()` becomes false.
7. `write_avail()` just returns 8192 and do not allocate new block